### PR TITLE
Implemented DI of scope provider in LoggerFactory

### LINKED ISSUE #50590: Allow replacing IExternalScopeProvider via DI
_Copied from https://github.com/dotnet/runtime/issues/52153 created by @Sundow:_

## Background and Motivation

Currently Microsoft.Extensions.Logging.LoggerFactory class always instantiates  IExternalScopeProvider with a default implementation. #50590 proposes to enable dependency injection so it is possible to use a custom implementation. This requires to add a constructor with one more parameter of type IExternalScopeProvider to be used by DI engine.

## Proposed API

```csharp
public partial class LoggerFactory : Microsoft.Extensions.Logging.ILoggerFactory
{
    public LoggerFactory(System.Collections.Generic.IEnumerable<Microsoft.Extensions.Logging.ILoggerProvider> providers, Microsoft.Extensions.Options.IOptionsMonitor<Microsoft.Extensions.Logging.LoggerFilterOptions> filterOption, Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Logging.LoggerFactoryOptions> options = null, Microsoft.Extensions.Logging.IExternalScopeProvider scopeProvider = null) { }
}
```

## Usage Examples

Will be implicitly used by DI engine.

## Risks

Low, as no existing interfaces are changed and only a trivial field assignment is required for a new one.

## Other notes
The actual implementation of the proposal already exists in #52132 and can be looked at there, but requires API change approval.

---

_Original description by @shirhatti:_

## Original Description

Today the default LoggerFactory implementation instantiates it's own IExternalScopeProvider. We should resolve it from DI instead allowing folks to replace the default IExternalScopeProvider.

### PR DESCRIPTION
**Pull request by Sundow**
Added LoggerFactory constructor with IExternalScopeProvider parameter allowing to use injected scope provider implementation. If nothing is provided uses the default implementation as before.

This change required an API change approval which is now done (see #50590).

Fix #50590

### CODE REVIEW

**Comment by KalleOlaviNiemitalo** [comment]
As in <https://github.com/dotnet/runtime/pull/52132#issuecomment-830370446>, I'd like a test that asserts that the LoggerFactory calls some method of the IExternalScopeProvider that was registered with DI.

**Comment by Sundow** [comment]
> As in [#52132 (comment)](https://github.com/dotnet/runtime/pull/52132#issuecomment-830370446), I'd like a test that asserts that the LoggerFactory calls some method of the IExternalScopeProvider that was registered with DI.

In general I agree with it and I was looking for a way to do it when developing a test. It just looks for me that developing such a test may require a way bigger effort that making the change itself as the whole infrastructure around needs to be mocked to test it.

Btw, I see that mono tests are failing on the committed test which didn't fail before. Something about mock libraries. Will look at it...

**Comment by KalleOlaviNiemitalo** [comment]
I think testing the feature would need:

- A class that implements IExternalScopeProvider and tracks whether Push has been called. It could forward to LoggerExternalScopeProvider instead of implementing from scratch.
- A class that implements ILoggerProvider and ISupportExternalScope. Could derive from NullLoggerProvider.
- Register those in the service collection and build the container.
- Resolve ILogger\<T>.
- Call ILogger.BeginScope and assert that IExternalScopeProvider.Push was called.\
  (I don't think ForEachScope would need to be tested separately, because LoggerFactory has only one IExternalScopeProvider field.)

Perhaps it is indeed a bit much.

**Review by phocodes — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Logging/tests/Common/LoggingServiceCollectionExtensionsTest.cs [hunk 1]
```diff
@@ -23,6 +24,22 @@ public void AddLogging_WrapsServiceCollection()
             Assert.True(callbackCalled);
         }
 
+        [Fact]
+        public void AddLogging_InjectScopeProvider()
+        {
+            var services = new ServiceCollection();
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging/tests/Common/LoggingServiceCollectionExtensionsTest.cs:30]: this test needs to be improved, needs a more relevant assertion.

**Comment by ericstj** [comment]
@Sundow can you address @maryamariyan's feedback?

**Comment by Sundow** [comment]
> @Sundow can you address @maryamariyan's feedback?

Trying. There are currently issues with building Microsoft.Extensions.Logging in Visual Studio. One of them (#53434) I fixed myself, it was easy. But another one (#54740) is harder for me because I don't know the build system well. Someone else is working on it. Not being able to build it in VS means I can't debug unit tests and I need it for the changes as they are not trivial.

**Comment by safern** [comment]
> Not being able to build it in VS means I can't debug unit tests and I need it for the changes as they are not trivial.

@Sundow I'm sorry you are hitting issues while building in VS. In the meantime you can try the following to debug the tests:

Add
```cs
while(!Debugger.IsAttached) { Thread.Sleep(100); }
Debugger.Break();
```

And then run the tests on the console and attach to the process from VS.

Another alternative without code changes is to open the `testhost/dotnet.exe` as a project in VS and then set the command line arguments that we print in the console when running the tests that say `"To reproduce: ..."`, and then open the files you want to set breakpoints on and execute it.

**Comment by maryamariyan** [comment]
The remaining 6.0 previews don't take new API additions. This would need to wait until 7.0. 

Thanks @Sundow for your contribution here. I'll close this and we could later reopen a new one, once the repo is ready for API additions again.

**Pull request closed without merging**